File: haushaltsbuch/src/main/java/org/github/haschi/haushaltsbuch/api/_Schulden.java
package org.github.haschi.haushaltsbuch.api;

import org.github.haschi.haushaltsbuch.infrastruktur.modellierung.de.Eingehüllt;
import org.github.haschi.haushaltsbuch.infrastruktur.modellierung.de.Umhüller;
import org.immutables.value.Value;
import org.javamoney.moneta.function.MonetaryFunctions;

import java.util.Collections;
import java.util.List;

@Value.Immutable
@Eingehüllt
public abstract class _Schulden extends Umhüller<List<Schuld>>
{
    public Währungsbetrag summe()
    {
        return Währungsbetrag.of(
                wert().stream()
                        .map(m -> m.währungsbetrag().wert())
                        .reduce(MonetaryFunctions.sum())
                        .orElse(Währungsbetrag.NullEuro().wert()));
    }

    public static Schulden leer()
    {
        return Schulden.of(Collections.emptyList());
    }
}

File: haushaltsbuch/src/main/kotlin/org/github/haschi/haushaltsbuch/core/Schulden.kt
package org.github.haschi.haushaltsbuch.core

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonValue
import org.github.haschi.haushaltsbuch.api.Währungsbetrag
import org.javamoney.moneta.function.MonetaryFunctions
import javax.money.MonetaryAmount

data class Schulden @JsonCreator constructor (@get:JsonIgnore private val l: List<Schuld>) : List<Schuld> by l {

    constructor(vararg ls: Schuld): this(ls.asList())

    @JsonValue
    fun wert(): List<Schuld> = l

    val summe: Währungsbetrag
    get() = Währungsbetrag(
            this.stream()
                    .map<MonetaryAmount> { m -> m.währungsbetrag.wert }
                    .reduce(MonetaryFunctions.sum())
                    .orElse(Währungsbetrag.NullEuro.wert))

    companion object {
        val keine: Schulden
            get() = Schulden(emptyList())
    }
}

